### PR TITLE
Add file and path manipulation utilities

### PR DESCRIPTION
**Pull request by smowton**
This upstreams a bunch of file and pathname manipulation utils we'd found useful in the security product. A fair few of them were rewritten just now by me to account for corner cases, so there are *very likely* bugs (hence the do not merge tag). Rather I'm putting this up now to see if people want them; then if so I'll write some unit tests and hence fix them up.

### CODE REVIEW

File: src/goto-analyzer/unreachable_instructions.cpp [hunk 1]
```diff
@@ -90,7 +90,7 @@ static void add_to_json(
 
   entry["function"]=json_stringt(id2string(end_function->function));
   entry["fileName"]=
-    json_stringt(concat_dir_file(
+    json_stringt(fileutil_concat_dir_file(
```
Comment (tautschnig) [inline src/goto-analyzer/unreachable_instructions.cpp:93]: I'm routinely allergic to intrusive changes: would it be possible to skip the renaming? If ambiguity really turns out to be a problem, a class (with static members) should be introduced.

File: src/util/file_util.cpp [hunk 2]
```diff
@@ -24,25 +24,34 @@ Date: January 2012
...
+#include <cstdlib>
+#include <algorithm>
+#include <sstream>
+#include <vector>
+
 #include <unistd.h>
 #include <dirent.h>
-#include <cstdlib>
 #include <cstdio>
+#include <sys/stat.h>
```
Comment (tautschnig) [inline src/util/file_util.cpp:37]: Moving this include looks like a rather random change?

File: src/util/file_util.cpp [hunk 3]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+# endif
+}
+
+/// Gets the size of an existing file.
+/// \param filename: file to query
+/// \return size in bytes
+/// \throws fileutil_error on error
+uint64_t fileutil_file_size(const std::string &filename)
+{
+  std::ifstream f{filename, std::ios::binary};
```
Comment (tautschnig) [inline src/util/file_util.cpp:196]: How about using `stat(2)`?

**Review by tautschnig — COMMENTED** [review]
I'm curious to learn what the use case for all these functions is. As they deal with user-controlled input, a lot of security concerns arise.

File: src/util/file_util.cpp [hunk 4]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+  // Full match of shorter_path?
+  if(mismatch_iterators.first==shorter_path.end() &&
+     *mismatch_iterators.second==PATH_SEPARATOR)
+    return shorter_path;
+
+  // Mismatch in shorter_path. Strip last path component:
+  std::string matching_part=
+    shorter_path.substr(
+      0, std::distance(shorter_path.begin(), mismatch_iterators.first));
+
```
Comment (LAJW) [inline src/util/file_util.cpp:374]: I think this whole chunk could be replaced by:
```c++
	return std::string(
		path1.begin(),
		std::mismatch(
			path1.begin(),
			path1.begin() + std::min(path1.size(), path2.size()),
			path2.begin()).first);

```

**Review by LAJW — APPROVED** [review]
Looks useful (with exception of the new file reading utility, why not just use `fstream`?). I'd prefer returning optional over throwing exceptions with unnecessarily detailed error messages.

I just have one nit-pick that I hope you might consider.

File: src/util/file_util.cpp (hunk 2 above)
Comment (LAJW) [reply src/util/file_util.cpp:37]: I'm not a fan of creating dedicated refactor PRs just to change one line. If such small piggybacked adjustments are discouraged, the code quality is bound to deteriorate over time.

Comment (tautschnig) [reply src/util/file_util.cpp:37]: I do agree, but I can't seem to make sense of this change? Why is this include better placed down here than where it used to sit?

File: src/util/file_util.cpp [hunk 5]
```diff
@@ -24,25 +24,34 @@ Date: January 2012
...
+#undef NOMINMAX
 #include <util/unicode.h>
 #define chdir _chdir
 #define popen _popen
 #define pclose _pclose
 #endif
 
 /// \return current working directory
-std::string get_current_working_directory()
+std::string fileutil_get_current_working_directory()
```
Comment (tautschnig) [inline src/util/file_util.cpp:54]: In light of the recent hiccup in the value-set analysis: can we please not do arbitrary renaming? Thanks.

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Comment by smowton** [comment]
@tautschnig re: use-cases, again this is from the security tool, which consumes and produces small filesystem trees (`result_dir/report.json`, `result_dir/scenario_1/binary.gbf`, ...). You can see how an implementation of `mkdir -p`, `path_join` and so on would be useful for this. Normally obviously I'd use some library, probably boost-filesystem (or wait for c++17 which finally brings an FS lib), but that's not allowed in cbmc. 

No worries if they're not wanted, just thought I'd see whether we can save other cbmc-derived products some time and effort since we've got these already.

**Comment by tautschnig** [comment]
> No worries if they're not wanted, just thought I'd see whether we can save other cbmc-derived products some time and effort since we've got these already.

I'm not suggesting this isn't wanted: 1) do not unnecessarily break existing code, i.e., don't rename functions without a deprecation path unless strictly required; 2) do whatever is possible to avoid introducing security holes (there is a lot of string manipulation going on, and thread safety is likely not guaranteed in many cases -- at least flag that).

**Comment by smowton** [comment]
Sure, I'll add deprecated renames for the existing utils. I figured named like `get_current_directory` are very generic and so kind of ambiguous about what library they belong to. Thread safety is of course not provided with functions like `file_exists` but I will explicitly note that. The safety question is trickier, since obviously I flatter myself that I can see every unescaped shell character and equally obviously I'm wrong. The only solution is proper battle-tested library code. Would you be willing to bat for copy-pasting some of boost-filesystem in here or similar?

**Comment by tautschnig** [comment]
> I figured named like `get_current_directory` are very generic and so kind of ambiguous about what library they belong to.

I do agree, but a careful, managed migration will be appreciated by users. Thank you for doing the extra work!

**Comment by smowton** [comment]
Ping @thk123 @allredj

File: src/util/file_util.cpp (hunk 2 above)
Comment (allredj) [reply src/util/file_util.cpp:37]: I agree with @tautschnig that this should not be changed, unless there is a good reason to it.

File: src/util/file_util.cpp [hunk 6]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+bool fileutil_is_file(const std::string &pathname)
+{
+# if defined(WIN32)
+  return PathFileExists(pathname.c_str()) && !PathIsDirectory(pathname.c_str());
+# elif defined(__linux__) || defined(__APPLE__)
+  struct stat buf;
+  int ret=stat(pathname.c_str(), &buf);
+  return !ret && S_ISREG(buf.st_mode);
+# else
+#   error "Unsuported platform."
```
Comment (allredj) [inline src/util/file_util.cpp:170]: `Unsuported` -> `Unsupported`

File: src/util/file_util.cpp [hunk 7]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+bool fileutil_is_directory(const std::string &pathname)
+{
+# if defined(WIN32)
+  return PathIsDirectory(pathname.c_str());
+# elif defined(__linux__) || defined(__APPLE__)
+  struct stat buf;
+  int ret=stat(pathname.c_str(), &buf);
+  return !ret && S_ISDIR(buf.st_mode);
+# else
+#   error "Unsuported platform."
```
Comment (allredj) [inline src/util/file_util.cpp:186]: `Unsuported` -> `Unsupported`

File: src/util/file_util.cpp (hunk 3 above)
Comment (allredj) [reply src/util/file_util.cpp:196]: Sorry for being so ignorant, but isn't `stat(2)` linux-specific?
About the code here, why don't we use `std::ios::ate` to go right to the end of the file without having to `seekg` from the start?

File: src/util/file_util.cpp [hunk 8]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+  const std::streampos end=f.tellg();
+  if(end==-1)
+    throw fileutil_error("Couldn't read end position of "+filename);
+  return end-begin;
+}
+
+/// Finds last path element offset
+/// \param path: path to find path separators in
+/// \return offset of the start of the last element in path, or zero if empty.
+static std::string::size_type last_path_element(const std::string &path)
```
Comment (allredj) [inline src/util/file_util.cpp:214]: The method name doesn't really convey its functionality.

File: src/util/file_util.cpp [hunk 9]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+{
+  auto found=path.rfind(PATH_SEPARATOR, path.size()-1);
+  return found==std::string::npos ? 0 : found+1;
+}
+
+/// Path to filename.
+/// \param path: path to extract filename from
+/// \return part of path after the last path separator, or all of path if none
+std::string fileutil_path_filename(const std::string &path)
```
Comment (allredj) [inline src/util/file_util.cpp:223]: Would rename to `fileutil_path_to_filename`.

File: src/util/file_util.cpp [hunk 10]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+{
+  return path.substr(last_path_element(path));
+}
+
+/// Path to directory name.
+/// \param path: path to extract directory name from
+/// \return part of path after the last path separator, or all of path if none
+std::string fileutil_path_dirname(const std::string &path)
```
Comment (allredj) [inline src/util/file_util.cpp:231]: Would rename to `fileutil_path_to_dirname`

File: src/util/file_util.cpp [hunk 11]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+  }
+  return ostr.str();
+}
+
+/// Get a common prefix for path1 and path2 (e.g. `/x/y/z` and `/x/yw` have
+/// common prefix `/x`). Note paths are not canonicalised first.
+/// \param path1: first path
+/// \param path2: second path
+/// \return common prefix
+std::string fileutil_get_common_prefix(
```
Comment (allredj) [inline src/util/file_util.cpp:347]: Looks like a good candidate for a few unit tests.

File: src/util/file_util.cpp [hunk 12]
```diff
@@ -145,3 +154,268 @@ std::string concat_dir_file(
...
+
+  return fileutil_path_dirname(matching_part);
+}
+
+/// Expresses a path relative to a directory if it is a prefix, or returns it
+/// unchanged otherwise. Note neither is canonicalised first.
+/// \param path: path to strip of prefix
+/// \param directory: directory prefix to remove if found
+/// \return relative path if applicable or passed path otherwise.
+std::string fileutil_get_relative_path(
```
Comment (allredj) [inline src/util/file_util.cpp:383]: This could have some unit tests too.

File: src/util/file_util.cpp (hunk 7 above)
Comment (allredj) [reply src/util/file_util.cpp:186]: Some of the methods here return `error`, whereas some throw an exception. Should it not be homogeneous?

**Review by allredj — COMMENTED** [review]
A few non-blocking comments.

**Comment by smowton** [comment]
@tautschnig aware this is stalled, going to replace this with a library, but it's low priority stuff. I'll kill this PR til then.

**Pull request closed without merging**